File: src/main/java/org/tud/oas/routing/IRoutingProvider.java
package org.tud.oas.routing;

import java.util.List;

import org.tud.oas.demand.IDemandView;
import org.tud.oas.supply.ISupplyView;

public interface IRoutingProvider {
    void setProfile(String profile);

    void setRangeType(String range_type);

    void setParameter(String name, Object value);

    /**
     * Computes the time-distance-matrix containing travel-times between all
     * supply and demand points.
     * Points that cannot be reached (either because of seperated sub-networks or
     * max search range will get a value < 0)
     * 
     * @param demand
     * @param supply
     * @param options
     * @return
     */
    ITDMatrix requestTDMatrix(IDemandView demand, ISupplyView supply, RoutingOptions options) throws Exception;

    /**
     * Computes the nearest supply to the demand points.
     * 
     * @param demand
     * @param supply
     * @param options
     * @return
     */
    INNTable requestNearest(IDemandView demand, ISupplyView supply, RoutingOptions options) throws Exception;

    /**
     * Computes the k-nearest supplies to the demand points.
     * 
     * @param demand
     * @param supply
     * @param k
     * @param options
     * @return
     */
    IKNNTable requestKNearest(IDemandView demand, ISupplyView supply, int k, RoutingOptions options) throws Exception;

    /**
     * Computes all supplies that lie within a given range to the demand
     * points.
     * 
     * @param demand
     * @param supply
     * @param range
     * @param options
     * @return
     */
    ICatchment requestCatchment(IDemandView demand, ISupplyView supply, double range, RoutingOptions options)
            throws Exception;
}

File: src/routing/IRoutingProvider.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Threading.Tasks.Dataflow;
using DVAN.Population;

namespace DVAN.Routing
{
    public interface IRoutingProvider
    {
        void setProfile(string profile);
        void setRangeType(string range_type);
        void setOption(string name, object value);

        /// <summary>
        /// Computes the time-distance-matrix containing travel-times between all facilities and population points.
        /// </summary>
        Task<ITDMatrix?> requestTDMatrix(IPopulationView population, double[][] facilities, List<double> ranges, string mode);

        /// <summary>
        /// Computes the nearest facilities to the population points.
        /// </summary>
        Task<INNTable?> requestNearest(IPopulationView population, double[][] facilities, List<double> ranges, string mode);

        /// <summary>
        /// Computes the k-nearest facilities to the population points.
        /// </summary>
        Task<IKNNTable?> requestKNearest(IPopulationView population, double[][] facilities, List<double> ranges, int k, string mode);


        /// <summary>
        /// Computes all facilities that lie within a given range to the population points.
        /// </summary>
        Task<ICatchment?> requestCatchment(IPopulationView population, double[][] facilities, double range, string mode);
    }
}